### PR TITLE
Fix a regression in encoder-decoder generation cache initialization

### PR DESCRIPTION
**Pull request by kaixuanliu**
For encoder-decoder models, `generate()` was passing the decoder config to both the self-attention cache and the cross-attention cache. This is incorrect for models like T5Gemma with decoder sliding-window layers: the cross-attention cache could inherit the decoder sliding-window structure and truncate encoder key/value states, causing FlashAttention generation to crash.

This PR keeps the decoder config for the self-attention cache, but initializes the cross-attention cache without the decoder config so cross-attention remains full-length.

Tested with:

```bash
pytest -q -rA tests/models/t5gemma/test_modeling_t5gemma.py::T5GemmaModelTest::test_generate_beyond_sliding_window_with_flash_attn
```
@vasqu @ArthurZucker @CyrilVallez, pls help review, thx!

### CODE REVIEW

**Comment by kaixuanliu** [comment]
Move the fix to t5gemma modeling part, and fix device mismatch bug for failed case:
`tests/models/t5gemma/test_modeling_t5gemma.py::T5GemmaModelTest::test_model_parallel_beam_search`

File: src/transformers/models/t5gemma/modular_t5gemma.py [hunk 1]
```diff
@@ -614,6 +614,16 @@ def forward(
             # We do not pass the config to the cross attn cache to avoid initializing SWA
             # --> we use full attention between our cross attentions
             past_key_values = EncoderDecoderCache(DynamicCache(config=self.config), DynamicCache())
+        elif (
```
Comment (vasqu) [inline src/transformers/models/t5gemma/modular_t5gemma.py]: Imo, we should override similar to t5gemma2 instead see https://github.com/huggingface/transformers/blob/e65c3a2d1461e6c43f6f5d4157c583846f63f71d/src/transformers/models/t5gemma2/modular_t5gemma2.py#L1106

File: src/transformers/models/t5gemma/modular_t5gemma.py [hunk 2]
```diff
@@ -637,12 +647,15 @@ def forward(
             }
 
         if not isinstance(cross_attn_mask_mapping := encoder_attention_mask, dict):
+            if encoder_attention_mask is None:
+                encoder_attention_mask = torch.ones(
+                    encoder_hidden_states.shape[:2], device=inputs_embeds.device, dtype=torch.bool
+                )
```
Comment (vasqu) [inline src/transformers/models/t5gemma/modular_t5gemma.py]: This seems weird to me, any reason this was actually needed? If yes, please add a comment as to why

**Review by vasqu — COMMENTED** [review]
Can we also add a fast test for this? Or at least a faster test?

Left some smaller comments

File: src/transformers/models/t5gemma/modular_t5gemma.py (hunk 2 above)
Comment (kaixuanliu) [reply src/transformers/models/t5gemma/modular_t5gemma.py]: Well, it is somewhat like a WA here. As if we pass `encoder_attention_mask` to `create_bidirectional_mask`, it will hit `early_exit` in [create_bidirectional_mask](https://github.com/huggingface/transformers/blob/main/src/transformers/masking_utils.py#L1057), and avoid the device mismatch issue afterwards. I think it should be a common issue for cross-attention mask, will fix the bug in masking_utils.py

File: src/transformers/models/t5gemma/modular_t5gemma.py (hunk 1 above)
Comment (kaixuanliu) [reply src/transformers/models/t5gemma/modular_t5gemma.py]: Done

**Comment by kaixuanliu** [comment]
@vasqu I have resolved the comments you left, can you help review it again? Thx!

File: src/transformers/masking_utils.py [hunk 3]
```diff
@@ -1058,7 +1058,11 @@ def create_bidirectional_mask(
         return attention_mask
 
     embeds = encoder_hidden_states if encoder_hidden_states is not None else inputs_embeds
-    batch_size, dtype, device = embeds.shape[0], embeds.dtype, embeds.device
+    batch_size, dtype = embeds.shape[0], embeds.dtype
+    # Use `inputs_embeds.device` to stay consistent with `_preprocess_mask_arguments`, which moves the 2D
+    # `attention_mask` to that device. In model parallel setups, `encoder_hidden_states` may live on a different
+    # device than `inputs_embeds` (e.g. cross-attention from a decoder to encoder states).
+    device = inputs_embeds.device
```
Comment (vasqu) [inline src/transformers/masking_utils.py]: Can we move this to a separate PR but thanks for fixing, that is a good point

**Review by vasqu — APPROVED** [review]
LGTM, can we move the mask fix to a different PR. Other than that we can merge then

**Comment by kaixuanliu** [comment]
@vasqu Done.

File: src/transformers/masking_utils.py (hunk 3 above)
Comment (kaixuanliu) [reply src/transformers/masking_utils.py]: Move to https://github.com/huggingface/transformers/pull/46221.

**Comment by vasqu** [comment]
Thanks! Merging
